fl4health/privacy_mechanisms/dpPoissonBinomial.py

```diff
@@ -0,0 +1,162 @@
+    Args:
+        query_vector (List[float]): Client integer vector.
+        radius (float): The radius.
+        tight_frame (List[float]): Matrix to compute Kashin representation.
+        theta (float): A parameter within the real interval [0, 1/4].
+        K (float): Constant parameter in Kashin representation, or clip.
+        modulus (int): The modulus for modular arithmetic used in SecAgg.
+
+    Raises:
+        Exception: Bad query.
+
+    Returns:
+        List[int]: The discretized and privatized query_vector as a list of integers.
+
+    Reference
+        The Poisson Binomial Mechanism for Unbiased Federated Learning with Secure Aggregation
+        https://proceedings.mlr.press/v162/chen22s/chen22s.pdf
+    """
+
+    dim = len(query_vector)
+
+    if dim == 0:
+        raise Exception("query_vector cannot be empty")
+
+    kashin_matrix = torch.tensor(tight_frame)
+    client_vector = torch.tensor(query_vector)
+
+    kashin_vector = torch.matmul(kashin_matrix, client_vector)
+
+    # TODO adjust this for l-infinity clipping
+    scalar_radius = radius * K / math.sqrt(dim)
+
+    privatization = torch.zeros(dim)
+    for j in range(dim):
+        privatization[j] += ScalerPoissonBinomial(kashin_vector[j].item(), modulus - 1, scalar_radius, theta)
+
+    return list(privatization)
+
+
+def PoissonBinomialMechanism_Server(
+    inverse_matrix: List[List[int]], radius: float, modulus: int, theta: float, *client_vectors: List[int]
+) -> List[int]:
+    """
+    Server side function that recovers the sum of vectors from clients.
+
+    Args:
+        inverse_matrix (List[List[int]]): The inverse of the random rotation generated by Kashin_alternative().
+        radius (float): Radius.
+        modulus (int): Modulus for modular arithmetic in SecAgg
+        theta (float): Parameter within the real interval [0, 1/4].
+
+    Returns:
+        List[int]: Mean estimator of the client updates.
+
+    Reference
+        Algorithm 1 in "The Poisson Binomial Mechanism for Unbiased Federated Learning with Secure Aggregation"
+        https://proceedings.mlr.press/v162/chen22s/chen22s.pdf
+    """
+
+    dim, num_clients = len(inverse_matrix), len(client_vectors)
+
+    # these assertions prevent division by zero
+    assert dim > 0 and num_clients > 0
+    assert 0 < theta <= 0.25
+
+    kashin_sum = torch.zeros(dim)  # Coordinates are in Kashin coordinates.
+
+    for vect in client_vectors:
+        kashin_sum += torch.tensor(vect)
+
+    kashin_sum -= modulus * num_clients / 2 * torch.ones(dim)
+    kashin_sum *= radius / (modulus * num_clients * theta)
+
+    rotation = torch.tensor(inverse_matrix)
+
+    estimator = torch.matmul(rotation, kashin_sum)
+
+    return estimator.tolist()
+
+
+if __name__ == "__main__":
+    # # ScalerPoissonBinomial test cases
+    # x, N, radius, theta = 0, 20, 5, 0.1
+    # out = [ScalerPoissonBinomial(x, N, radius, theta) for _ in range(100)]
+    # print(out)
+
+    # PoissonBinomialMechanism test cases
```

Comment:
I would recommend making this a formal test and also running tests on the various helper functions to ensure that they are working properly.